{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import packages \n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from statsmodels.stats.multitest import multipletests"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1) Import GO enrichement output (Rstudio)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import output from GO enrichment analysis in Rstudio \n",
    "enrichment = pd.read_csv('/home/maggy/WholeGenomeBisulphiteSequencing/GO_analysis/2_filter_GO_erichement_analysis/NvsA_GOterms.txt', sep='\\t')\n",
    "\n",
    "# Import subset (for caluclating corrections)\n",
    "subset = pd.read_csv('/home/maggy/WholeGenomeBisulphiteSequencing/GO_analysis/1_prepare_data_for_GO_analysis/goseq_NvsA.csv', sep=',', index_col=0)\n",
    "\n",
    "# import the gene lenth df to get the geneids for the background \n",
    "background = pd.read_csv('/home/maggy/WholeGenomeBisulphiteSequencing/GO_analysis/1_prepare_data_for_GO_analysis/gene_length.csv', sep=',')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# calculate expected number in each category \n",
    "enrichment['expectedInCat'] = (enrichment['numDEInCat']/50000)* 35\n",
    "# 50000 number of GO terms \n",
    "\n",
    "\n",
    "# calculate fold enrichment\n",
    "enrichment['fold_enrichment'] = enrichment['numDEInCat']/enrichment['expectedInCat']\n",
    "\n",
    "# calculate negative fold change from the FDR corrected values\n",
    "enrichment['-log10(FDR)'] = -np.log10(enrichment['over_represented_pvalue'])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "import statsmodels.stats.multitest as smm\n",
    "enrichment['corrected_p_values'] = smm.multipletests(enrichment['over_represented_pvalue'], method='bonferroni')[1]\n",
    "\n",
    "\n",
    "import statsmodels.stats.multitest as smm\n",
    "\n",
    "enrichment['q_values'] = smm.multipletests(enrichment['over_represented_pvalue'], method='fdr_bh')[1]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/maggy/anaconda3/envs/methylscore/lib/python3.11/site-packages/statsmodels/stats/multitest.py:186: RuntimeWarning: divide by zero encountered in log1p\n",
      "  np.log1p(-pvals))\n"
     ]
    }
   ],
   "source": [
    "import statsmodels.stats.multitest as smm\n",
    "\n",
    "enrichment['corrected_p_values'] = smm.multipletests(enrichment['over_represented_pvalue'], alpha=0.05, method='hs')[1]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>category</th>\n",
       "      <th>over_represented_pvalue</th>\n",
       "      <th>under_represented_pvalue</th>\n",
       "      <th>numDEInCat</th>\n",
       "      <th>numInCat</th>\n",
       "      <th>term</th>\n",
       "      <th>ontology</th>\n",
       "      <th>adjusted_pvalue</th>\n",
       "      <th>expectedInCat</th>\n",
       "      <th>fold_enrichment</th>\n",
       "      <th>-log10(FDR)</th>\n",
       "      <th>corrected_p_values</th>\n",
       "      <th>q_values</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GO:0061015</td>\n",
       "      <td>0.000863</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>snRNA import into nucleus</td>\n",
       "      <td>BP</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0007</td>\n",
       "      <td>1428.571429</td>\n",
       "      <td>3.063873</td>\n",
       "      <td>0.937040</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GO:0007018</td>\n",
       "      <td>0.001802</td>\n",
       "      <td>0.999965</td>\n",
       "      <td>2</td>\n",
       "      <td>141</td>\n",
       "      <td>microtubule-based movement</td>\n",
       "      <td>BP</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0014</td>\n",
       "      <td>1428.571429</td>\n",
       "      <td>2.744332</td>\n",
       "      <td>0.996887</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GO:0015272</td>\n",
       "      <td>0.002356</td>\n",
       "      <td>0.999998</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>ATP-activated inward rectifier potassium chann...</td>\n",
       "      <td>MF</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0007</td>\n",
       "      <td>1428.571429</td>\n",
       "      <td>2.627815</td>\n",
       "      <td>0.999473</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GO:0032543</td>\n",
       "      <td>0.004122</td>\n",
       "      <td>0.999993</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>mitochondrial translation</td>\n",
       "      <td>BP</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0007</td>\n",
       "      <td>1428.571429</td>\n",
       "      <td>2.384910</td>\n",
       "      <td>0.999998</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>GO:0030334</td>\n",
       "      <td>0.004379</td>\n",
       "      <td>0.999992</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>regulation of cell migration</td>\n",
       "      <td>BP</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0007</td>\n",
       "      <td>1428.571429</td>\n",
       "      <td>2.358594</td>\n",
       "      <td>0.999999</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3198</th>\n",
       "      <td>GO:0015936</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.997678</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>coenzyme A metabolic process</td>\n",
       "      <td>BP</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3199</th>\n",
       "      <td>GO:0006888</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.976066</td>\n",
       "      <td>0</td>\n",
       "      <td>55</td>\n",
       "      <td>endoplasmic reticulum to Golgi vesicle-mediate...</td>\n",
       "      <td>BP</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3200</th>\n",
       "      <td>GO:0061666</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999032</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>UFM1 ligase activity</td>\n",
       "      <td>MF</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3201</th>\n",
       "      <td>GO:0070573</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.997408</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>metallodipeptidase activity</td>\n",
       "      <td>MF</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3202</th>\n",
       "      <td>GO:1990542</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.998322</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>mitochondrial transmembrane transport</td>\n",
       "      <td>BP</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3202 rows Ã— 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        category  over_represented_pvalue  under_represented_pvalue   \n",
       "1     GO:0061015                 0.000863                  1.000000  \\\n",
       "2     GO:0007018                 0.001802                  0.999965   \n",
       "3     GO:0015272                 0.002356                  0.999998   \n",
       "4     GO:0032543                 0.004122                  0.999993   \n",
       "5     GO:0030334                 0.004379                  0.999992   \n",
       "...          ...                      ...                       ...   \n",
       "3198  GO:0015936                 1.000000                  0.997678   \n",
       "3199  GO:0006888                 1.000000                  0.976066   \n",
       "3200  GO:0061666                 1.000000                  0.999032   \n",
       "3201  GO:0070573                 1.000000                  0.997408   \n",
       "3202  GO:1990542                 1.000000                  0.998322   \n",
       "\n",
       "      numDEInCat  numInCat                                               term   \n",
       "1              1         2                          snRNA import into nucleus  \\\n",
       "2              2       141                         microtubule-based movement   \n",
       "3              1         5  ATP-activated inward rectifier potassium chann...   \n",
       "4              1         9                          mitochondrial translation   \n",
       "5              1        10                       regulation of cell migration   \n",
       "...          ...       ...                                                ...   \n",
       "3198           0         5                       coenzyme A metabolic process   \n",
       "3199           0        55  endoplasmic reticulum to Golgi vesicle-mediate...   \n",
       "3200           0         2                               UFM1 ligase activity   \n",
       "3201           0         6                        metallodipeptidase activity   \n",
       "3202           0         4              mitochondrial transmembrane transport   \n",
       "\n",
       "     ontology  adjusted_pvalue  expectedInCat  fold_enrichment  -log10(FDR)   \n",
       "1          BP                1         0.0007      1428.571429     3.063873  \\\n",
       "2          BP                1         0.0014      1428.571429     2.744332   \n",
       "3          MF                1         0.0007      1428.571429     2.627815   \n",
       "4          BP                1         0.0007      1428.571429     2.384910   \n",
       "5          BP                1         0.0007      1428.571429     2.358594   \n",
       "...       ...              ...            ...              ...          ...   \n",
       "3198       BP                1         0.0000              NaN    -0.000000   \n",
       "3199       BP                1         0.0000              NaN    -0.000000   \n",
       "3200       MF                1         0.0000              NaN    -0.000000   \n",
       "3201       MF                1         0.0000              NaN    -0.000000   \n",
       "3202       BP                1         0.0000              NaN    -0.000000   \n",
       "\n",
       "      corrected_p_values  q_values  \n",
       "1               0.937040       1.0  \n",
       "2               0.996887       1.0  \n",
       "3               0.999473       1.0  \n",
       "4               0.999998       1.0  \n",
       "5               0.999999       1.0  \n",
       "...                  ...       ...  \n",
       "3198            1.000000       1.0  \n",
       "3199            1.000000       1.0  \n",
       "3200            1.000000       1.0  \n",
       "3201            1.000000       1.0  \n",
       "3202            1.000000       1.0  \n",
       "\n",
       "[3202 rows x 13 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "enrichment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "module 'matplotlib.pyplot' has no attribute 'dot'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[7], line 5\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mmatplotlib\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mpyplot\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m \u001b[38;5;21;01mplt\u001b[39;00m\n\u001b[1;32m      3\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mseaborn\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m \u001b[38;5;21;01msns\u001b[39;00m\n\u001b[0;32m----> 5\u001b[0m \u001b[43mplt\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdot\u001b[49m(enrichment[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mfold_enrichment\u001b[39m\u001b[38;5;124m'\u001b[39m], enrichment[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m-log10(FDR)\u001b[39m\u001b[38;5;124m'\u001b[39m], alpha\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0.5\u001b[39m)\n",
      "\u001b[0;31mAttributeError\u001b[0m: module 'matplotlib.pyplot' has no attribute 'dot'"
     ]
    }
   ],
   "source": [
    "# PLOT \n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "sns.scatterplot(data=enrichment, x='-log10(FDR)', y='fold_enrichment', hue='over_represented_pvalue')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "methylscore",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
