{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load packages \n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# import packages \n",
    "import re\n",
    "import shutil\n",
    "import os\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1 Import total number of genes "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import total number of genes from DESeq2 code \n",
    "total = pd.read_csv(\"12_gene_length_distribution/total_number_of_genes.csv\")\n",
    "total = total.rename(columns={\"x\": \"gene_name\"})\n",
    "\n",
    "# should be 45667 genes\n",
    "#len(total[\"gene_name\"].unique())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2 Import list of gene names \n",
    "with more that 10 reads in total (over all three conditions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import genes with more than 10 reads as sum() from DESeq2 code \n",
    "genenames = pd.read_csv(\"12_gene_length_distribution/number_of_genes_with_total_of_10_reads.csv\")\n",
    "genenames = genenames.rename(columns={\"x\": \"gene_name\"})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2 Import annotation   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import carcar_annotation_v5.gtf \n",
    "annotation = pd.read_csv('12_gene_length_distribution/carcar_annotation_v5.gtf', sep='\\t', comment='#', header=None)\n",
    "# renaming the header of the annotation file\n",
    "annotation = annotation.rename({0: 'seqname',\n",
    "                                1: 'source',\n",
    "                                2: 'feature',\n",
    "                                3: 'start',\n",
    "                                4: 'stop',\n",
    "                                5: 'score',\n",
    "                                6: 'strand',\n",
    "                                7: 'frame',\n",
    "                                8: 'attribute'}, axis=1)\n",
    "\n",
    "    \n",
    "# define a function to extract the geneid\n",
    "def extract_gene_id(attributes):\n",
    "    # Using a regular expression to match and extract the desired part\n",
    "    match = re.search(r'gene_id \"([^\"]+)\"', attributes)\n",
    "    if match:\n",
    "        return match.group(1)\n",
    "    else:\n",
    "        return 0 # Return 0 if there is no match\n",
    "    \n",
    "# filtering the column feature for transcritps\n",
    "annotation = annotation[annotation['feature'] == 'transcript']\n",
    "# Apply the function to the \"attributes\" column and create a new column with the extracted gene ids\n",
    "annotation['gene_id'] = annotation['attribute'].apply(extract_gene_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>scaffold_1</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>transcript</td>\n",
       "      <td>246717</td>\n",
       "      <td>248723</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"scaffold_1-g45517\"; transcript_id \"sc...</td>\n",
       "      <td>scaffold_1-g45517</td>\n",
       "      <td>2006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>scaffold_1</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>transcript</td>\n",
       "      <td>316526</td>\n",
       "      <td>317083</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"scaffold_1-g45518\"; transcript_id \"sc...</td>\n",
       "      <td>scaffold_1-g45518</td>\n",
       "      <td>557</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>scaffold_1</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>transcript</td>\n",
       "      <td>316526</td>\n",
       "      <td>317116</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"scaffold_1-g45518\"; transcript_id \"sc...</td>\n",
       "      <td>scaffold_1-g45518</td>\n",
       "      <td>590</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>scaffold_1</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>transcript</td>\n",
       "      <td>317117</td>\n",
       "      <td>317526</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"scaffold_1-g45519\"; transcript_id \"sc...</td>\n",
       "      <td>scaffold_1-g45519</td>\n",
       "      <td>409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>scaffold_1</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>transcript</td>\n",
       "      <td>319660</td>\n",
       "      <td>320224</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"scaffold_1-g45520\"; transcript_id \"sc...</td>\n",
       "      <td>scaffold_1-g45520</td>\n",
       "      <td>564</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2161996</th>\n",
       "      <td>scaffold_72</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>transcript</td>\n",
       "      <td>43108</td>\n",
       "      <td>43483</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"scaffold_72-g45760\"; transcript_id \"s...</td>\n",
       "      <td>scaffold_72-g45760</td>\n",
       "      <td>375</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2162002</th>\n",
       "      <td>scaffold_86</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>transcript</td>\n",
       "      <td>17392</td>\n",
       "      <td>31005</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"scaffold_86-g45770\"; transcript_id \"s...</td>\n",
       "      <td>scaffold_86-g45770</td>\n",
       "      <td>13613</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2162020</th>\n",
       "      <td>scaffold_129</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>transcript</td>\n",
       "      <td>346</td>\n",
       "      <td>2361</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"scaffold_129-g45565\"; transcript_id \"...</td>\n",
       "      <td>scaffold_129-g45565</td>\n",
       "      <td>2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2162028</th>\n",
       "      <td>scaffold_129</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>transcript</td>\n",
       "      <td>2573</td>\n",
       "      <td>9987</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"scaffold_129-g45566\"; transcript_id \"...</td>\n",
       "      <td>scaffold_129-g45566</td>\n",
       "      <td>7414</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2162040</th>\n",
       "      <td>scaffold_186</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>transcript</td>\n",
       "      <td>2678</td>\n",
       "      <td>6006</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"scaffold_186-g45580\"; transcript_id \"...</td>\n",
       "      <td>scaffold_186-g45580</td>\n",
       "      <td>3328</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>82557 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              seqname    source     feature   start    stop score strand  \\\n",
       "1          scaffold_1  AUGUSTUS  transcript  246717  248723     .      -   \n",
       "9          scaffold_1  AUGUSTUS  transcript  316526  317083     .      -   \n",
       "12         scaffold_1  AUGUSTUS  transcript  316526  317116     .      -   \n",
       "16         scaffold_1  AUGUSTUS  transcript  317117  317526     .      -   \n",
       "22         scaffold_1  AUGUSTUS  transcript  319660  320224     .      -   \n",
       "...               ...       ...         ...     ...     ...   ...    ...   \n",
       "2161996   scaffold_72  AUGUSTUS  transcript   43108   43483     .      -   \n",
       "2162002   scaffold_86  AUGUSTUS  transcript   17392   31005     .      +   \n",
       "2162020  scaffold_129  AUGUSTUS  transcript     346    2361     .      -   \n",
       "2162028  scaffold_129  AUGUSTUS  transcript    2573    9987     .      -   \n",
       "2162040  scaffold_186  AUGUSTUS  transcript    2678    6006     .      -   \n",
       "\n",
       "        frame                                          attribute  \\\n",
       "1           .  gene_id \"scaffold_1-g45517\"; transcript_id \"sc...   \n",
       "9           .  gene_id \"scaffold_1-g45518\"; transcript_id \"sc...   \n",
       "12          .  gene_id \"scaffold_1-g45518\"; transcript_id \"sc...   \n",
       "16          .  gene_id \"scaffold_1-g45519\"; transcript_id \"sc...   \n",
       "22          .  gene_id \"scaffold_1-g45520\"; transcript_id \"sc...   \n",
       "...       ...                                                ...   \n",
       "2161996     .  gene_id \"scaffold_72-g45760\"; transcript_id \"s...   \n",
       "2162002     .  gene_id \"scaffold_86-g45770\"; transcript_id \"s...   \n",
       "2162020     .  gene_id \"scaffold_129-g45565\"; transcript_id \"...   \n",
       "2162028     .  gene_id \"scaffold_129-g45566\"; transcript_id \"...   \n",
       "2162040     .  gene_id \"scaffold_186-g45580\"; transcript_id \"...   \n",
       "\n",
       "                     gene_id  gene_length  \n",
       "1          scaffold_1-g45517         2006  \n",
       "9          scaffold_1-g45518          557  \n",
       "12         scaffold_1-g45518          590  \n",
       "16         scaffold_1-g45519          409  \n",
       "22         scaffold_1-g45520          564  \n",
       "...                      ...          ...  \n",
       "2161996   scaffold_72-g45760          375  \n",
       "2162002   scaffold_86-g45770        13613  \n",
       "2162020  scaffold_129-g45565         2015  \n",
       "2162028  scaffold_129-g45566         7414  \n",
       "2162040  scaffold_186-g45580         3328  \n",
       "\n",
       "[82557 rows x 11 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3 Calculate gene length "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# calculate the gene length\n",
    "annotation['gene_length'] = annotation['stop'] - annotation['start']\n",
    "# export gene length as csv \n",
    "annotation[['gene_id', 'gene_length']].sort_values(by='gene_length', ascending=True).drop_duplicates(subset='gene_id', keep='last').to_csv(\"12_gene_length_distribution/ccar_gene_length.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1           2006\n",
       "9            557\n",
       "12           590\n",
       "16           409\n",
       "22           564\n",
       "           ...  \n",
       "2161996      375\n",
       "2162002    13613\n",
       "2162020     2015\n",
       "2162028     7414\n",
       "2162040     3328\n",
       "Name: gene_length, Length: 82557, dtype: int64"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotation['gene_length']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# keep only the lines of the annotation df that contain the gene_id from the gene names list \n",
    "data = pd.merge(annotation, genenames, left_on='gene_id', right_on='gene_name')\n",
    "# take total? "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# which genes are not unique \n",
    "data['gene_name'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# calculate max length of gene length\n",
    "data['length'] = data['stop'] - data['start']\n",
    "maxima = data['length'].max()\n",
    "minima = data['length'].min()\n",
    "# calculate the average\n",
    "mean_genelength = data['length'].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# PLOT: Length Distribution of Genes \n",
    "plt.figure(figsize=(10, 6))\n",
    "plt.title('Length Distribution of genes')\n",
    "# on the x axis: Length of genes\n",
    "plt.xlabel('Length of genes')\n",
    "# on the y axis: Number of genes\n",
    "plt.ylabel('Number of genes')\n",
    "\n",
    "# plot the histogram\n",
    "plt.hist(data['stop'] - data['start'], bins=range(minima,maxima, 10000), color='#B2D6B1', edgecolor='black', linewidth=0.05)\n",
    "plt.yscale('log')\n",
    "\n",
    "\n",
    "\n",
    "# LABELS: \n",
    "# add label for total number of genes \n",
    "plt.text(0.95, 0.95, f'Total number of genes:\\n{len(total[\"gene_name\"])}', ha='right', va='top', transform=plt.gca().transAxes)\n",
    "# add total number of unique genes \n",
    "plt.text(0.95, 0.85, f\"Genes with ≥ 10 reads:\\n{len(data['gene_name'].unique())}\", ha='right', va='top', transform=plt.gca().transAxes)\n",
    "# add numer label for average at the top right corner \n",
    "plt.text(0.95, 0.75, f'Average gene length:\\n{mean_genelength:.1f}', ha='right', va='top', transform=plt.gca().transAxes)\n",
    "\n",
    "plt.text(0.95, 0.65, f'Gene length rage:\\n{minima} - {maxima}', ha='right', va='top', transform=plt.gca().transAxes)\n",
    "\n",
    "\n",
    "# save the plot as png with 500 dpi\n",
    "plt.savefig('12_gene_length_distribution/gene_length_distribution.png', dpi=500)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "rnaseq",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
