{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import packages \n",
    "import pandas as pd \n",
    "import numpy as np \n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from pathlib import Path\n",
    "import statsmodels.stats.multitest as smm\n",
    "import json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import GO data \n",
    "# Geneid to GO term\n",
    "goterm = pd.read_csv('11_GO_enrichment_analysis/ccar2swissprotGO_long.txt', sep='\\t') \n",
    "# GO term to description\n",
    "godesc = pd.read_csv('11_GO_enrichment_analysis/go_data.txt', sep='\\t')\n",
    "godesc.set_index('id', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import DMR (for calculating the expectedDEInCat value) \n",
    "# I need the length of the DMRs \n",
    "with open(f'/home/maggy/WholeGenomeBisulphiteSequencing/GO_analysis/1_prepare_data_for_GO_analysis/dmr_unique_geneids.json', 'r') as file:\n",
    "    dmr_geneids = json.load(file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "go_files = sorted(Path(\"11_GO_enrichment_analysis\").glob('*_GOterms.csv'))\n",
    "\n",
    "# total number of genes in the background dataset \n",
    "total_num_genes = pd.read_csv('/home/maggy/WholeGenomeBisulphiteSequencing/RNAseq_pipeline/12_gene_length_distribution/total_number_of_genes.csv', sep=',')\n",
    "\n",
    "# create empty dict for df's\n",
    "go_enriched = []\n",
    "\n",
    "\n",
    "for f in go_files: \n",
    "    go_terms = pd.read_csv(f, sep=',')\n",
    "    # strip emtpy space around GO terms \n",
    "    go_terms['category'] = go_terms['category'].str.strip()\n",
    "    # since the term and ontology column are empty, I filled them with information from the godesc dataframe\n",
    "    go_terms['term'] = godesc.loc[go_terms['category'], 'name'].to_numpy()\n",
    "    go_terms['ontology'] = godesc.loc[go_terms['category'], 'namespace'].to_numpy()\n",
    "    # calculate expected number in each category \n",
    "    go_terms['expectedDEInCat'] = (go_terms['numInCat']/len(total_num_genes))*len(dmr_geneids[f.stem[:4]])\n",
    "    # calculate fold enrichment\n",
    "    go_terms['foldEnrichment'] = (go_terms['numDEInCat']/go_terms['expectedDEInCat'])\n",
    "    # calculate negative fold change from the FDR corrected values\n",
    "    go_terms['-log10(FDR)'] = -np.log10(go_terms['over_represented_pvalue'])\n",
    "    go_terms['padj'] = smm.multipletests(go_terms['over_represented_pvalue'], method='bonferroni')[1]\n",
    "    go_terms = go_terms.sort_values(by='padj')\n",
    "    #go_terms = go_terms[go_terms['padj'] <= 0.05]#.iloc[-10:]\n",
    "    if 'NvsA' in f.name: \n",
    "        go_terms['comparison'] = 'Normoxia to Anoxia'\n",
    "    elif 'NvsR' in f.name:\n",
    "        go_terms['comparison'] = 'Normoxia to Reoxygenation'\n",
    "    elif 'AvsR' in f.name:\n",
    "        go_terms['comparison'] = 'Anoxia to Reoxygenation'\n",
    "    \n",
    "    # save to list \n",
    "    go_enriched.append(go_terms)\n",
    "# merge the dataframes and add column with name of comparison\n",
    "go = pd.concat(go_enriched)\n",
    "go['logpadj'] = np.log10(go['padj'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "dmr_test = go.loc[go['over_represented_pvalue']<= 0.05]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>category</th>\n",
       "      <th>over_represented_pvalue</th>\n",
       "      <th>under_represented_pvalue</th>\n",
       "      <th>numDEInCat</th>\n",
       "      <th>numInCat</th>\n",
       "      <th>term</th>\n",
       "      <th>ontology</th>\n",
       "      <th>expectedDEInCat</th>\n",
       "      <th>foldEnrichment</th>\n",
       "      <th>-log10(FDR)</th>\n",
       "      <th>padj</th>\n",
       "      <th>comparison</th>\n",
       "      <th>logpadj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GO:0036042</td>\n",
       "      <td>0.000632</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>long-chain fatty acyl-CoA binding</td>\n",
       "      <td>MF</td>\n",
       "      <td>0.000657</td>\n",
       "      <td>1522.233333</td>\n",
       "      <td>3.199491</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Anoxia to Reoxygenation</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>GO:0120293</td>\n",
       "      <td>0.009760</td>\n",
       "      <td>0.999957</td>\n",
       "      <td>1</td>\n",
       "      <td>33</td>\n",
       "      <td>dynein axonemal particle</td>\n",
       "      <td>CC</td>\n",
       "      <td>0.010839</td>\n",
       "      <td>92.256566</td>\n",
       "      <td>2.010551</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Anoxia to Reoxygenation</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>GO:0021954</td>\n",
       "      <td>0.009663</td>\n",
       "      <td>0.999958</td>\n",
       "      <td>1</td>\n",
       "      <td>32</td>\n",
       "      <td>central nervous system neuron development</td>\n",
       "      <td>BP</td>\n",
       "      <td>0.010511</td>\n",
       "      <td>95.139583</td>\n",
       "      <td>2.014872</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Anoxia to Reoxygenation</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>GO:0016242</td>\n",
       "      <td>0.009480</td>\n",
       "      <td>0.999960</td>\n",
       "      <td>1</td>\n",
       "      <td>30</td>\n",
       "      <td>negative regulation of macroautophagy</td>\n",
       "      <td>BP</td>\n",
       "      <td>0.009854</td>\n",
       "      <td>101.482222</td>\n",
       "      <td>2.023179</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Anoxia to Reoxygenation</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>GO:0007155</td>\n",
       "      <td>0.009372</td>\n",
       "      <td>0.999960</td>\n",
       "      <td>1</td>\n",
       "      <td>30</td>\n",
       "      <td>cell adhesion</td>\n",
       "      <td>BP</td>\n",
       "      <td>0.009854</td>\n",
       "      <td>101.482222</td>\n",
       "      <td>2.028177</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Anoxia to Reoxygenation</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>GO:0070034</td>\n",
       "      <td>0.016985</td>\n",
       "      <td>0.999865</td>\n",
       "      <td>1</td>\n",
       "      <td>29</td>\n",
       "      <td>telomerase RNA binding</td>\n",
       "      <td>MF</td>\n",
       "      <td>0.017781</td>\n",
       "      <td>56.240148</td>\n",
       "      <td>1.769943</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Normoxia to Reoxygenation</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>GO:0007141</td>\n",
       "      <td>0.016978</td>\n",
       "      <td>0.999865</td>\n",
       "      <td>1</td>\n",
       "      <td>29</td>\n",
       "      <td>male meiosis I</td>\n",
       "      <td>BP</td>\n",
       "      <td>0.017781</td>\n",
       "      <td>56.240148</td>\n",
       "      <td>1.770111</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Normoxia to Reoxygenation</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>GO:0019725</td>\n",
       "      <td>0.016439</td>\n",
       "      <td>0.999874</td>\n",
       "      <td>1</td>\n",
       "      <td>28</td>\n",
       "      <td>cellular homeostasis</td>\n",
       "      <td>BP</td>\n",
       "      <td>0.017168</td>\n",
       "      <td>58.248724</td>\n",
       "      <td>1.784128</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Normoxia to Reoxygenation</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>GO:0043069</td>\n",
       "      <td>0.016427</td>\n",
       "      <td>0.999874</td>\n",
       "      <td>1</td>\n",
       "      <td>28</td>\n",
       "      <td>negative regulation of programmed cell death</td>\n",
       "      <td>BP</td>\n",
       "      <td>0.017168</td>\n",
       "      <td>58.248724</td>\n",
       "      <td>1.784441</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Normoxia to Reoxygenation</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>GO:0007342</td>\n",
       "      <td>0.015851</td>\n",
       "      <td>0.999883</td>\n",
       "      <td>1</td>\n",
       "      <td>27</td>\n",
       "      <td>fusion of sperm to egg plasma membrane involve...</td>\n",
       "      <td>BP</td>\n",
       "      <td>0.016555</td>\n",
       "      <td>60.406085</td>\n",
       "      <td>1.799951</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Normoxia to Reoxygenation</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>418 rows Ã— 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      category  over_represented_pvalue  under_represented_pvalue  numDEInCat   \n",
       "1   GO:0036042                 0.000632                  1.000000           1  \\\n",
       "53  GO:0120293                 0.009760                  0.999957           1   \n",
       "52  GO:0021954                 0.009663                  0.999958           1   \n",
       "51  GO:0016242                 0.009480                  0.999960           1   \n",
       "50  GO:0007155                 0.009372                  0.999960           1   \n",
       "..         ...                      ...                       ...         ...   \n",
       "99  GO:0070034                 0.016985                  0.999865           1   \n",
       "98  GO:0007141                 0.016978                  0.999865           1   \n",
       "97  GO:0019725                 0.016439                  0.999874           1   \n",
       "96  GO:0043069                 0.016427                  0.999874           1   \n",
       "94  GO:0007342                 0.015851                  0.999883           1   \n",
       "\n",
       "    numInCat                                               term ontology   \n",
       "1          2                  long-chain fatty acyl-CoA binding       MF  \\\n",
       "53        33                           dynein axonemal particle       CC   \n",
       "52        32          central nervous system neuron development       BP   \n",
       "51        30              negative regulation of macroautophagy       BP   \n",
       "50        30                                      cell adhesion       BP   \n",
       "..       ...                                                ...      ...   \n",
       "99        29                             telomerase RNA binding       MF   \n",
       "98        29                                     male meiosis I       BP   \n",
       "97        28                               cellular homeostasis       BP   \n",
       "96        28       negative regulation of programmed cell death       BP   \n",
       "94        27  fusion of sperm to egg plasma membrane involve...       BP   \n",
       "\n",
       "    expectedDEInCat  foldEnrichment  -log10(FDR)  padj   \n",
       "1          0.000657     1522.233333     3.199491   1.0  \\\n",
       "53         0.010839       92.256566     2.010551   1.0   \n",
       "52         0.010511       95.139583     2.014872   1.0   \n",
       "51         0.009854      101.482222     2.023179   1.0   \n",
       "50         0.009854      101.482222     2.028177   1.0   \n",
       "..              ...             ...          ...   ...   \n",
       "99         0.017781       56.240148     1.769943   1.0   \n",
       "98         0.017781       56.240148     1.770111   1.0   \n",
       "97         0.017168       58.248724     1.784128   1.0   \n",
       "96         0.017168       58.248724     1.784441   1.0   \n",
       "94         0.016555       60.406085     1.799951   1.0   \n",
       "\n",
       "                   comparison  logpadj  \n",
       "1     Anoxia to Reoxygenation      0.0  \n",
       "53    Anoxia to Reoxygenation      0.0  \n",
       "52    Anoxia to Reoxygenation      0.0  \n",
       "51    Anoxia to Reoxygenation      0.0  \n",
       "50    Anoxia to Reoxygenation      0.0  \n",
       "..                        ...      ...  \n",
       "99  Normoxia to Reoxygenation      0.0  \n",
       "98  Normoxia to Reoxygenation      0.0  \n",
       "97  Normoxia to Reoxygenation      0.0  \n",
       "96  Normoxia to Reoxygenation      0.0  \n",
       "94  Normoxia to Reoxygenation      0.0  \n",
       "\n",
       "[418 rows x 13 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dmr_test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "dmr_test_na = dmr_test.loc[dmr_test['comparison'] == 'Normoxia to Anoxia']\t\n",
    "dmr_test_nr = dmr_test.loc[dmr_test['comparison'] == 'Normoxia to Reoxygenation']\n",
    "dmr_test_ar = dmr_test.loc[dmr_test['comparison'] == 'Anoxia to Reoxygenation']\n",
    "\n",
    "# select only two columns\n",
    "dmr_test_na = dmr_test_na[['category', 'over_represented_pvalue']]\n",
    "dmr_test_nr = dmr_test_nr[['category', 'over_represented_pvalue']]\n",
    "dmr_test_ar = dmr_test_ar[['category', 'over_represented_pvalue']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "dmr_test_na.to_csv('11_GO_enrichment_analysis/2_GO_NvsA.csv', index=False, sep='\\t')\n",
    "dmr_test_nr.to_csv('11_GO_enrichment_analysis/2_GO_NvsR.csv', index=False, sep='\\t')\n",
    "dmr_test_ar.to_csv('11_GO_enrichment_analysis/2_GO_AvsR.csv', index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "methylscore",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
